x/margin/keeper/keeper_test.go

```diff
@@ -0,0 +1,585 @@
+
+	if !reflect.DeepEqual(got, want) {
+		t.Errorf("got %v want %v", got, want)
+	}
+}
+
+func TestKeeper_DestroyMTP(t *testing.T) {
+	t.Run("key does not exist", func(t *testing.T) {
+		ctx, _, marginKeeper := initKeeper(t)
+		got := marginKeeper.DestroyMTP(ctx, "ceth", "xxx")
+
+		assertError(t, got, types.ErrMTPDoesNotExist)
+	})
+	t.Run("key exists", func(t *testing.T) {
+		ctx, app, marginKeeper := initKeeper(t)
+		addMTPKey(t, ctx, app, marginKeeper, "ceth", "xxx")
+		got := marginKeeper.DestroyMTP(ctx, "ceth", "xxx")
+
+		assertNoError(t, got)
+	})
+}
+
+func TestKeeper_ClpKeeper(t *testing.T) {
+	_, _, marginKeeper := initKeeper(t)
+	marginKeeper.ClpKeeper()
+}
+
+func TestKeeper_BankKeeper(t *testing.T) {
+	_, _, marginKeeper := initKeeper(t)
+	marginKeeper.BankKeeper()
+}
+
+func TestKeeper_GetLeverageParam(t *testing.T) {
+	ctx, _, marginKeeper := initKeeper(t)
+	marginKeeper.GetLeverageParam(ctx)
+}
+
+func TestKeeper_CustodySwap(t *testing.T) {
+	asset := clptypes.Asset{Symbol: "rowan"}
+	pool := clptypes.Pool{
+		ExternalAsset:        &asset,
+		NativeAssetBalance:   sdk.NewUint(1000000000),
+		NativeLiabilities:    sdk.NewUint(1000000000),
+		ExternalCustody:      sdk.NewUint(1000000000),
+		ExternalAssetBalance: sdk.NewUint(1000000000),
+		ExternalLiabilities:  sdk.NewUint(1000000000),
+		NativeCustody:        sdk.NewUint(1000000000),
+		PoolUnits:            sdk.NewUint(1),
+		Health:               sdk.NewDec(1),
+	}
+
+	custodySwapTests := []struct {
+		name       string
+		denom      string
+		decimals   int64
+		to         string
+		sentAmount sdk.Uint
+		err        error
+	}{
+		{
+			name:       "denom not registered",
+			denom:      "unregistred_denom",
+			decimals:   18,
+			to:         "xxx",
+			sentAmount: sdk.NewUint(0),
+			err:        tokenregistrytypes.ErrNotFound,
+		},
+		{
+			name:       "invalid sent amount",
+			denom:      "rowan",
+			decimals:   18,
+			to:         "xxx",
+			sentAmount: sdk.NewUint(0),
+			err:        nil,
+		},
+		{
+			name:       "no token adjustment and non-rowan target asset",
+			denom:      "rowan",
+			decimals:   18,
+			to:         "xxx",
+			sentAmount: sdk.NewUint(10000000000000),
+			err:        clptypes.ErrNotEnoughAssetTokens,
+		},
+		{
+			name:       "no token adjustment and rowan target asset",
+			denom:      "rowan",
+			decimals:   18,
+			to:         "rowan",
+			sentAmount: sdk.NewUint(10000000000000),
+			err:        clptypes.ErrNotEnoughAssetTokens,
+		},
+		{
+			name:       "token adjustment and non-rowan target asset",
+			denom:      "rowan",
+			decimals:   9,
+			to:         "xxx",
+			sentAmount: sdk.NewUint(1000),
+			err:        nil,
+		},
+		{
+			name:       "token adjustment and rowan target asset",
+			denom:      "rowan",
+			decimals:   9,
+			to:         "rowan",
+			sentAmount: sdk.NewUint(1000000000000),
+			err:        clptypes.ErrNotEnoughAssetTokens,
+		},
+	}
+
+	for _, tt := range custodySwapTests {
+		t.Run(tt.name, func(t *testing.T) {
```

Comment:
For the linter errors below, insert the line tt := tt before passing tt to t.Run()